### PR TITLE
ci: Only run CI jobs if needed

### PR DESCRIPTION
**Pull request by alexec**
Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this is a chore.
* [x] The title of the PR is (a) [conventional](https://www.conventionalcommits.org/en/v1.0.0/), (b) states what changed, and (c) suffixes the related issues number. E.g. `"fix(controller): Updates such and such. Fixes #1234"`.  
* [x] I have written unit and/or e2e tests for my change. PRs without these are unlike to be merged.
* [x] Optional. I've added My organization is added to the USERS.md.
* [x] I've signed the CLA and required builds are green. 

# Goals

The goal of this PR is to make CI builds faster. 

It does this by skipping jobs using `circleci step halt` to skip jobs. Jobs are run the diffs between the master and the branch contain changes to files that could result in the job failing.

### CODE REVIEW

File: util/tmp/tmp.go [hunk 1]
```diff
@@ -0,0 +1 @@
+package tmp
```
Comment (alexec) [inline util/tmp/tmp.go:1]: these tmp files are intended to make sure jobs do actually run when needed

File: hack/maybe-skip-job.sh [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+[ "$branch" = master ] && exit
+
+# tip - must use origin/master for CircleCI
+diffs=$(git diff --name-only origin/master)
+
+# if there are changes to this areas, we must run
+rx=
+case $job in
+codegen)
+  rx='api/\|manifests/\|pkg/'
```
Comment (simster7) [inline hack/maybe-skip-job.sh:17]: This doesn't seem exhaustive. I can think of other files that can affect codegen:

- Changes in `hack/`
- Changes in some `.go` files that are mocked

Comment (alexec) [reply hack/maybe-skip-job.sh:17]: Adding them is easy - so I'll add hack. I'm not sure .go files you mean?

Comment (alexec) [reply hack/maybe-skip-job.sh]: oh - this would include all go files in 'pkg/'

Comment (simster7) [reply hack/maybe-skip-job.sh]: I had in mind something like this: `persist/sqldb/mocks/OffloadNodeStatusRepo.go`. I.e., files under `*/mock/*`

Comment (alexec) [reply hack/maybe-skip-job.sh]: we'll need to patch this I expect as we find problems

**Pull request merged**